src/PersistentSettingsStore.js

```diff
@@ -0,0 +1,185 @@
+'use strict'
+
+let equal = require('deep-equal');
+let deepcopy = require('deepcopy');
+
+import { logger, configureLogger } from './logger';
+let authDataManager = require('./authDataManager');
+let DatabaseAdapter = require('./DatabaseAdapter');
+
+// doesn't make sense to expose / modify these settings
+let lockedSettings = [
+  'applicationId',
+  'masterKey',
+  'serverURL',
+  'collectionPrefix',
+  'filesController',
+  'pushController',
+  'loggerController',
+  'hooksController',
+  'userController',
+  'authDataManager',
+  'liveQueryController'
+];
+
+// callbacks for specific setting changes
+let onChange = {
+  logLevel: logLevel => {
+    configureLogger({ level: logLevel });
+    return logLevel;
+  },
+  oauth: (oauth, doc) => {
+    doc.locked.authDataManager = authDataManager(oauth, doc.settings.enableAnonymousUsers);
+    return oauth;
+  },
+  enableAnonymousUsers: (enableAnonymousUsers, doc) => {
+    doc.locked.authDataManager = authDataManager(doc.settings.oauth, enableAnonymousUsers);
+    return enableAnonymousUsers;
+  },
+  sessionLength: sessionLength => Number(sessionLength)
+}
+
+export default function PersistentSettingsStore(options, definedSettings) {
+  let {
+    freshness = 15,
+    lockDefinedSettings = true
+  } = options;
+  let dataStore = {};
+
+  return {
+    // filter can be 'locked' or 'persisted'
+    get: (key, filter) => {
+      let doc = dataStore[key];
+      pullSettings(doc);
+      if (filter) {
+        // if filter, return wrapped settings for locked / persisted
+        return filterDescriptors(doc.settings, doc[filter]);
+      } else {
+        // return wrapped settings
+        return doc.settings;
+      }
+    },
+
+    set: (key, settings) => {
+      let doc = {
+        // An object with getter/setters wrapped properties. Values stored in persisted/locked
+        settings: settings,
+        // Store for the persisted properties of settings
+        persisted: {},
+        // Store for the locked properties of settings
+        locked: {},
+        // Last time the settings were pulled from the database
+        lastPull: new Date(1970, 0, 0),
+      };
+
+      // wrap settings with getters / setters
+      setupSettingPersistence(doc, settings);
+
+      // place doc in map
+      dataStore[key] = doc;
+
+      // sync with database
+      return pullSettings(doc).then(_ => pushSettings(doc));
+    },
+
+    remove: key => {
+      delete dataStore[key];
+    },
+
+    clear: _ => {
+      dataStore = {};
+    }
+  };
+
+  function getSettingsCollection(doc) {
+    return DatabaseAdapter.getDatabaseConnection(doc.settings.applicationId, doc.settings.collectionPrefix)
+      .adaptiveCollection('_ServerSettings');
```

Comment:
we wanna avoid direct access to the DB Adapter, so that would need to go through the DBController instead